lib/varnishcfg/backends.go

```diff
@@ -0,0 +1,172 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+func (v *VCLBuilder) configureDirectors(vclFile *vclFile, parents *atscfg.ParentAbstraction) ([]string, error) {
+	warnings := []string{}
+
+	vclFile.imports = append(vclFile.imports, "directors")
+	var err error
+	requestFQDNs := make([]string, 0)
+
+	for _, svc := range parents.Services {
+		addBackends(vclFile.backends, append(svc.Parents, svc.SecondaryParents...), svc.DestDomain, svc.Port)
+		addDirectors(vclFile.subroutines, svc)
+
+		requestFQDNs = []string{svc.DestDomain}
+
+		if v.toData.Server.Type == tc.CacheTypeEdge.String() {
+			dsRegexes := atscfg.MakeDSRegexMap(v.toData.DeliveryServiceRegexes)
+			anyCastPartners := atscfg.GetAnyCastPartners(v.toData.Server, v.toData.Servers)
+			requestFQDNs, err = atscfg.GetDSRequestFQDNs(
+				&svc.DS,
+				dsRegexes[tc.DeliveryServiceName(*svc.DS.XMLID)],
+				v.toData.Server,
+				anyCastPartners,
+				v.toData.CDN.DomainName,
+			)
+			if err != nil {
+				warnings = append(warnings, "error getting ds '"+*svc.DS.XMLID+"' request fqdns, skipping! Error: "+err.Error())
+				continue
+			}
+		}
+
+		assignBackends(vclFile.subroutines, svc, requestFQDNs)
+	}
+
+	return warnings, nil
+}
+
+func assignBackends(subroutines map[string][]string, svc *atscfg.ParentAbstractionService, requestFQDNs []string) {
+	lines := make([]string, 0)
+
+	conditions := make([]string, 0)
+	for _, fqdn := range requestFQDNs {
+		conditions = append(conditions, fmt.Sprintf(`req.http.host == "%s"`, fqdn))
+	}
+
+	lines = append(lines, fmt.Sprintf("if (%s) {", strings.Join(conditions, " || ")))
+	lines = append(lines, fmt.Sprintf("\tset req.backend_hint = %s.backend();", svc.Name))
+
+	// only change request host from edge servers which typically has multiple request FQDNs or
+	// one request FQDN that is not the origin.
+	if len(requestFQDNs) > 1 || (len(requestFQDNs) == 1 && requestFQDNs[0] != svc.DestDomain) {
+		lines = append(lines, fmt.Sprintf("\tset req.http.host = \"%s\";", svc.DestDomain))
+	}
+
+	lines = append(lines, "}")
+
+	if _, ok := subroutines["vcl_recv"]; !ok {
+		subroutines["vcl_recv"] = make([]string, 0)
+	}
+	subroutines["vcl_recv"] = append(subroutines["vcl_recv"], lines...)
+}
+
+func addBackends(backends map[string]backend, parents []*atscfg.ParentAbstractionServiceParent, originDomain string, originPort int) {
+	for _, parent := range parents {
+		backendName := fmt.Sprintf("%s", getBackendName(parent.FQDN, parent.Port))
+		if _, ok := backends[backendName]; ok {
+			continue
+		}
+		backends[backendName] = backend{
+			host: parent.FQDN,
+			port: parent.Port,
+		}
+	}
+	backendName := fmt.Sprintf("%s", getBackendName(originDomain, originPort))
+	if _, ok := backends[backendName]; ok {
+		return
```

Comment:
Why skip this silently? Seems like it would be an error if the parent and origin backend names conflict?

Reply:
while that is definitely an issue if in the same topology a parent had conflicts with the origin it can still happen across topologies as this is the map of all backends the server currently added. So, a warning here might report some cases which is not an issue.